src/main/scala/tech/sourced/gemini/Gemini.scala

```diff
@@ -81,39 +83,112 @@ class Gemini(session: SparkSession, log: Slf4jLogger, keyspace: String = Gemini.
-      val features = extractFeatures(feClient, uast.get)
-      log.info(s"features: ${features.toString}")
-      // TODO: calculate minhash, find similar
+  def bytes2hex(bytes: Array[Byte], sep: Option[String] = None): String = {
+    sep match {
+      case None => bytes.map("%02x".format(_)).mkString
+      case Some(s) => bytes.map("%02x".format(_)).mkString(s)
+    }
+  }
+
+  def findSimilarForFile(
+                          file: File,
+                          conn: Session,
+                          bblfshClient: BblfshClient,
+                          feClient: FeatureExtractor,
+                          docFreqFile: String,
+                          paramsFile: String,
+                          htnum: Int,
+                          bandSize: Int): Option[Array[String]] = {
+
+    extractUAST(file, bblfshClient) match {
+      case Some(uast: Node) => {
+        log.debug(s"uast received: ${uast.toString}")
+        val features = extractFeatures(feClient, uast)
+        log.debug(s"features: ${features.toString}")
+        val wmh = hashFile(features, new File(docFreqFile), new File(paramsFile))
+
+        // FIXME try to rewrite later
+        var bands = Array.empty[Array[Byte]]
+        for (i <- 0 until htnum) {
+          var band = Array.empty[Byte]
+          for (j <- i * bandSize until (i + 1) * bandSize) {
+            var vArray = BigInt(wmh(j)(0)).toByteArray
+            for (_ <- vArray.size until 4) {
+              vArray = vArray :+ 0.toByte
+            }
+            vArray = vArray ++ BigInt(wmh(j)(1)).toByteArray
+            for (_ <- vArray.size until 8) {
+              vArray = vArray :+ 0.toByte
+            }
+
+            band = band ++ vArray
+          }
+          bands = bands :+ band
+        }
+
+        log.info("Looking for similar items")
+        val similar = bands.zipWithIndex.foldLeft(Set.empty[String]) { case (similar, (band, i)) =>
+          val table = "hashtables"
+          val cql = s"SELECT sha1 FROM $keyspace.$table WHERE hashtable=$i AND value=0x${bytes2hex(band)}"
+          log.debug(cql)
+
+          val sha1s = conn.execute(new SimpleStatement(cql))
+            .asScala
+            .map(row => row.getString("sha1"))
+
+          similar ++ sha1s
+        }
+        log.info(s"Fetched ${similar.size} items")
+
+        Some(similar.toArray)
+      }
+      case _ => None
     }
   }
 
   def extractUAST(file: File, client: BblfshClient): Option[Node] = {
     val byteArray = Files.readAllBytes(file.toPath)
 
-    val resp = client.parse(file.getName, new String(byteArray))
-    if (resp.errors.nonEmpty) {
-      val errors = resp.errors.mkString(",")
-      log.error(s"bblfsh errors: ${errors}")
-    }
+    try {
+      val resp = client.parse(file.getName, new String(byteArray))
+      if (resp.errors.nonEmpty) {
+        val errors = resp.errors.mkString(",")
+        log.error(s"bblfsh errors: ${errors}")
+      }
 
-    resp.uast
+      resp.uast
```

Comment:
Not related to the changes, but I think it's safer to return `None` in case there were any `resp.errors` present as well.

Reply:
I return uast based on how bblfsh works.

LA team can't confirm which drivers work like this right now, but the original idea was:
- bblfsh can parse partly incorrect code (for example if there is error only in one function, this function would be skipped but the rest of code will be parsed correctly)
- such errors go to `resp.errors`

Though I agree it's arguable that we need partial parsing in gemini or not.
Please confirm that we shouldn't return uast in case of any error taken in account info above.

Reply:
Thank you for the followup!

> LA team can't confirm which drivers work like this right now, but the original idea was:

That is the thing.. the goal for us though is just to be consistent between hash/query. 

So any of the ways would do for now, it's just that in my previous experience working with bblfsh, it was safer and more reproducible to fail on any error.

Reply:
yes. I couldn't find code that checks for errors in src-d ml. So to be consistent with hash I also return partial UAST. We can change it later when we implement hash on our side.

thanks for heads up though!